### PR TITLE
Retry pthread_kill on Linux if result is EAGAIN

### PR DESCRIPTION
**Pull request by gbalykov**
This PR enables `pthread_kill` retry on undocumented `EAGAIN` return value. This fix is similar to the one from mono: https://github.com/dotnet/runtime/pull/33966. 

More detailed description: https://github.com/dotnet/runtime/issues/32377#issuecomment-602734256. One more related discussion with example: https://lkml.org/lkml/2009/3/18/61.

We probably observe `EAGAIN` returned from `pthread_kill` with next backtrace on coreclr 3.1:
```
#0  __GI_raise (sig=sig@entry=6) at /usr/src/debug/glibc-2.30/signal/../sysdeps/unix/sysv/linux/raise.c:50
#1  0xb605ac20 in __GI_abort () at /usr/src/debug/glibc-2.30/stdlib/abort.c:79
#2  0xaf0af058 in PROCAbort () at src/pal/src/thread/process.cpp:3485
#3  0xaf08a100 in InjectActivationInternal (pThread=<optimized out>) at src/pal/src/exception/signal.cpp:697
#4  0xaf0b28a6 in PAL_InjectActivation (hThread=<optimized out>) at src/pal/src/thread/thread.cpp:2952
#5  0xaeebe69a in Thread::InjectGcSuspension (this=<optimized out>) at src/vm/threadsuspend.cpp:6774
#6  ThreadSuspend::SuspendRuntime (reason=<optimized out>) at src/vm/threadsuspend.cpp:4263
#7  0xaeec0776 in ThreadSuspend::SuspendEE (reason=ThreadSuspend::SUSPEND_FOR_GC) at src/vm/threadsuspend.cpp:6514
#8  0xaefd35d0 in GCToEEInterface::SuspendEE (reason=(unknown: 0)) at src/vm/gcenv.ee.cpp:25
#9  0xaefb30a8 in WKS::GCHeap::GarbageCollectGeneration (this=<optimized out>, gen=<optimized out>, reason=<optimized out>) at src/gc/gc.cpp:36545
#10 0xaefb4f78 in WKS::gc_heap::try_allocate_more_space (acontext=<optimized out>, size=524, flags=<optimized out>, gen_number=0) at src/gc/gc.cpp:13934
#11 0xaefce488 in WKS::gc_heap::allocate_more_space (flags=0, alloc_generation_number=0, acontext=<optimized out>, size=<optimized out>) at src/gc/gc.cpp:14369
#12 WKS::gc_heap::allocate (jsize=<optimized out>, acontext=<optimized out>, flags=<optimized out>) at src/gc/gc.cpp:14400
#13 WKS::GCHeap::Alloc (this=<optimized out>, context=0x1563460, size=524, flags=0) at src/gc/gc.cpp:35827
#14 0xaee6c8d8 in AllocateSzArray (pArrayMT=0x31f3e02c, cElements=256, flags=<optimized out>, bAllocateInLargeHeap=1) at src/vm/gchelpers.cpp:240
#15 0xaee7790c in JIT_NewArr1 (arrayMT=0x31f3e02c, size=256) at src/vm/jithelpers.cpp:3145
#16 0x31be3b82 in ?? ()
Backtrace stopped: previous frame identical to this frame (corrupt stack?)
```

`EAGAIN` is currently (.net 6, not coreclr 3.1) considered as normal ret code in `InjectActivationInternal` because of possible stack overflow in one thread during stack overflow processing for another thread (https://github.com/dotnet/runtime/pull/32167). But after this PR `EAGAIN` retried for 5 times will be considered an error and abort will happen, and I think this PR breaks double stack overflow case. What can be a possible workaround for `EAGAIN` returned during `GCToEEInterface::SuspendEE` when signal queue is full on linux?

cc @alpencolt @jkotas @janvorli

### CODE REVIEW

File: src/coreclr/pal/src/exception/signal.cpp [hunk 1]
```diff
@@ -838,21 +838,53 @@ static void inject_activation_handler(int code, siginfo_t *siginfo, void *contex
...
+    {
+        status = pthread_kill(pThread->GetPThreadSelf(), INJECT_ACTIVATION_SIGNAL);
+        // We can get EAGAIN when printing stack overflow stack trace and when other threads hit
+        // stack overflow too. Those are held in the sigsegv_handler with blocked signals until
+        // the process exits.
+        //
+        // We can also get EAGAIN if the signal queue overflows on linux.
+
+#ifdef __linux__
+        if (status == EAGAIN && retry_count < signal_queue_ovf_retry_count)
```
Comment (jkotas) [inline src/coreclr/pal/src/exception/signal.cpp:872]: I do not think we need to have a retry loop here. The caller of `InjectActivation` in the runtime has a retry loop already. I think we can just return failure from here - similar to how we return `ENOTSUP` on Apple - and let the runtime retry.

Comment (jkotas) [reply src/coreclr/pal/src/exception/signal.cpp:872]: Or maybe we should just delete `PROCAbort` on failure and propagate all error to runtime.

@janvorli Thoughts?

**Comment by janvorli** [comment]
I don't think we need to change anything here. The EAGAIN is checked and it is considered benign here. We don't abort on EAGAIN. The retry happens automatically in the suspension code no matter whether `InjectActivation` returns true or false.

File: src/coreclr/pal/src/exception/signal.cpp (hunk 1 above)
Comment (janvorli) [reply src/coreclr/pal/src/exception/signal.cpp:872]: We never check the return code of the `InjectActivation` for other purposes than logging a stresslog entry.

Comment (janvorli) [reply src/coreclr/pal/src/exception/signal.cpp:872]: We rely on the retry loop in the runtime. The `InjectActivation` can succeed and still the callback would never be called if the thread was executing a native code.

**Comment by janvorli** [comment]
@gbalykov when you mentioned double stack overflow, did you mean stack overflow during stack overflow handling? That is something that the runtime needs to make its best effort to avoid. The stack overflow handling code does variety of things to do that. E.g. executing the SIGSEGV on an alternate stack, allocating extra stack overflow handling stack on Unix or reserving enough extra stack pages on Windows.

**Comment by gbalykov** [comment]
> did you mean stack overflow during stack overflow handling?

I meant situation described in this comment: https://github.com/dotnet/runtime/pull/32167/files#diff-5306d97c9d437cb622c02d0dc3a4255b7ea63ace472fe625b0400f4806be3ab1R758-R761, i.e. when one thread handles stack overflow, and stack overflow also happens in the other one. That PR made `EAGAIN` valid ok return from `InjectActivationInternal`, and combined with retry on runtime side I conclude this should work fine in case of signal queue overflow on linux. We probably have observed exactly this behavior on coreclr 3.1, where `EAGAIN` change was not yet made, so I was curious what would be the correct way to fix it and whether .net6/master would also need any change. Thanks for clarification!

**Pull request closed without merging**